R/GeospatialFunctions.R

```diff
@@ -0,0 +1,611 @@
+      
+      return(list("TADA_with_ATTAINS" = no_ATTAINS_data,
+                  "ATTAINS_catchments" = ATTAINS_catchments,
+                  "ATTAINS_points" = ATTAINS_points,
+                  "ATTAINS_lines" = ATTAINS_lines,
+                  "ATTAINS_polygons" = ATTAINS_polygons))
+    } else {
+      return(no_ATTAINS_data)}
+    
+    # If there IS ATTAINS data...
+  } else {
+    
+    suppressMessages(suppressWarnings({
+      # ... link WQP features to the ATTAINS catchment feature(s) they land in:
+      TADA_with_ATTAINS <- TADA_DataRetrieval_data %>%
+        # left join = TRUE to preserve all observations (with or without ATTAINS features):
+        sf::st_join(., nearby_catchments, left = TRUE)
+      
+      if(return_sf == FALSE){
+        return(TADA_with_ATTAINS)
+      }
+      
+      # CATCHMENT FEATURES 
+      # use original catchment pull, but return column names to original
+      ATTAINS_catchments <- nearby_catchments
+      colnames(ATTAINS_catchments) <- gsub("ATTAINS.", "", colnames(ATTAINS_catchments)) 
+      # due to the rename, must re-set geometry column:
+      sf::st_geometry(ATTAINS_catchments) <- "geometry"
+      
+      # POINT FEATURES - try to pull point AU data if it exists. Otherwise, move on...
+      ATTAINS_points <- NULL
+      try(ATTAINS_points <- attains_features[["ATTAINS_points"]] %>%
+            # subset to only ATTAINS point features in the same NHD HR catchments as WQP observations
+            .[nearby_catchments,] %>%
+            # make sure no duplicate features exist
+            dplyr::distinct(assessmentunitidentifier, .keep_all = TRUE),
+          silent = TRUE)
+      
+      # LINE FEATURES - try to pull line AU data if it exists. Otherwise, move on...
+      ATTAINS_lines <- NULL
+      try(ATTAINS_lines <- attains_features[["ATTAINS_lines"]] %>%
+            # subset to only ATTAINS line features in the same NHD HR catchments as WQP observations
+            .[nearby_catchments,] %>%
+            # make sure no duplicate line features exist
+            dplyr::distinct(assessmentunitidentifier, .keep_all = TRUE),
+          silent = TRUE)
+      
+      # POLYGON FEATURES - try to pull polygon AU data if it exists. Otherwise, move on...
+      ATTAINS_polygons <- NULL
+      try(ATTAINS_polygons <- attains_features[["ATTAINS_polygons"]] %>%
+            # subset to only ATTAINS polygon features in the same NHD HR catchments as WQP observations
+            .[nearby_catchments,] %>%
+            # make sure no duplicate polygon features exist
+            dplyr::distinct(assessmentunitidentifier, .keep_all = TRUE),
+          silent = TRUE)
+    }))
+    
+    return(list("TADA_with_ATTAINS" = TADA_with_ATTAINS,
+                "ATTAINS_catchments" = ATTAINS_catchments,
+                "ATTAINS_points" = ATTAINS_points,
+                "ATTAINS_lines" = ATTAINS_lines,
+                "ATTAINS_polygons" = ATTAINS_polygons))
+  }
+  
+}
+
+
+#' TADA_ViewATTAINS
```

Comment:
I really like the appearance of the features in the map. It is easy to identify the lines/polygons/points vs the catchment outlines relative to the MonitoringLocations. I am wondering if long term it would make sense to pull catchment outlines from a source other than ATTAINS because my understanding is that only catchments with associated Assessment Units are available from the ATTAINS webservices. Being able to see catchment outlines in areas where an org has not yet created an assessment unit (but has new MonitoringLocations) might be very useful when making determinations about which MonitoringLocations should be included within new assessment units.

Reply:
I agree - "Being able to see catchment outlines in areas where an org has not yet created an assessment unit (but has new MonitoringLocations) might be very useful when making determinations about which MonitoringLocations should be included within new assessment units.". 

I'll write up a separate issue to capture this, and we can discuss details with the team (e.g. what additional sources for catchment we would want to use and how they match up/don't match up with the catchments ATTAINS uses) as a potential future improvement. For example, I believe the ATTAINS catchments are a mix of NHD high res catchments and hexagons for some oceans/lakes/estuaries. I am not sure how that will mesh visually with if we bring in a different catchment layer for the areas that the ATTAINS catchment layer does not cover.